### DISCUSSION TITLE
A few questions about functionality

### DISCUSSION THREAD

**Discussion by darynwhite** [predict-idlab/plotly-resampler#217] [Q&A] [answered]
Thanks for making this critically needed tool for the plotly API!

I have a few questions based on the attached GIF examples. Our use case is to view hi-resolution (1-minute to 10-minute sampling) environmental observations for post-deployment QC. We will be starting with an overview of the entire length of the deployment, typically 12-16 months, and need to be able to zoom into the granular level. In my testing I have discovered the attached behavior and have these questions about it:

- Is it possible to freeze the y-axis? It would greatly help during the review for the y-axis to remain the same
- Is the observed behavior when looking at the individual sample scale expected?
- I know this has been discussed, but "Reset Axis" doesn't cause a resample; is there a workaround with Dash?
- In the second example, the stacking of samples is confusing to observe and determine which sample is for which time, is this expected or a by-product of using a time-series?

![plotly-resampler-problem](https://github.com/predict-idlab/plotly-resampler/assets/40304663/91d0ee2a-65e6-40cc-89a2-28d904997fb8)

![plotly-resampler-problem-2](https://github.com/predict-idlab/plotly-resampler/assets/40304663/031081b6-2fcd-417e-aa95-a2e85221f2b5)

**Comment by jvdd** [comment]
Hey @darynwhite 

Thanks for sharing your questions here - we appreciate the effort you put into adding the GIFs :) 
I'll try to answer them one by one (and ask more detail where necessary).

**Q1 - Freezing the y-axis**
This is more a plotly related question (if I am understanding you correctly) concerning the figure layout. 
You can fixate the y-axis range - after resampling (e.g., zooming) the y-axis range does not change :arrow_down: 
```py
# fix the y-axis range
fig.layout.yaxis.range = [-2, 2]
```

**Q2 - behavior at the individual sample scale**
If you zoom in enough - narrowing the data range down to fewer than `max_n_samples` - you should be seeing a vanilla plotly plot.
I suppose you are referring to the weird / unexpected behavior displayed in the 2nd GIF when zoomed in, i.e., the points do not move when you are panning the x-axis...
I tried to reproduce this locally, but did not observe this behavior - can you perhaps share some minimal reproducible code in a dedicated issue?

**Q3 - "Reset Axis" does cause a resample**
Are you talking about the "Reset Axis" button (located at the right top of the figure its modebar)? Clicking this button resets the view (thereby resampling the data) to the original view?
I suspect you intend something else - feel free to further clarify this (or even provide some minimal code in a dedicated issue).

**Q4 - stacking of samples**
Not really sure whether I am properly understanding this question.
But here are my 5 cents;
- you can use a vertical line on hover: `fig.update_layout(hovermode="x unified")`
- you can play around with opacity to better discern different samples

Hope this helps! Please further comment if I misinterpreted any of your questions :)

Cheers, Jeroen

P.S.: Your use-case for plotly-resampler sound really fascinating! If you have any additional information / interesting plotly-resampler empowered visualizations that you can share with us - you may always contact us via mail :)

**Reply by darynwhite** [reply]
I'll add a quick update that I have on the second plot GIF: https://github.com/plotly/plotly.js/issues/3522

This is attached to a deep-level bug with the libraries used for GL plots on macOS. Apologies for thinking it was `plotly-resampler`!

**Reply by darynwhite** [reply]
Sure enough, it is that. I swapped to a `Scatter` trace and it functions slower, but how we expect it to without resampling.

_this is in reference to Q2 and Q4, both are caused the GL library issue on macOS_

**Reply by darynwhite** [reply, marked as answer]
More digging produced this issue thread that solved the stacking/slipping problem observed: https://github.com/plotly/plotly.js/issues/5970

**Comment by darynwhite** [comment]
Hey @jvdd!

I'm going to break Q1 out here. What's happening is that I set the `y-axis`'s `range` and `fixedrange` parameters in my plotting code, but when I plot with a `Resampler` object, the `y-axis` changes on zoom. You can see the behavior in both GIFs. I'm working on an example of the plotting code, but that will have to come later as I have it broken out into a library for our use case.

**Reply by darynwhite** [reply]
I'll address both problems I'm seeing, but hopefully, I'm applying things in the correct order.

Here is what I'm seeing:
![plotly-resampler-axes-problems](https://github.com/predict-idlab/plotly-resampler/assets/40304663/6a653c05-4efd-4303-9664-f830a4f047d2)

If you watch the `y-axis` labels, you'll see that it adjusts the values/range upon zooming and resampling. This is something I'm hoping to fix the entire time because it helps give context to where we are at in the deployment or seasonal cycle by seeing space above or below.

At the end of the GIF, you'll see that I use the "Reset Axes" button and it does reset but doesn't trip a resampling event. When I then zoom out and in with the graph buttons, a resample triggers and the graph fully resets. Maybe there is something in my setup that is causing this, but I'm unsure where to look for this problem.

Thanks for all of your work on this! I hope Plotly takes notice because this is a vital thing that they don't have anything remotely similar to what you have built.

**Comment by jvdd** [comment]
Hi @darynwhite, this is a fascinating environmental-QC use case. I have been building [FlexViz](https://github.com/flex-analytics/flexviz) (successor to this project) that is packed with much more functionality (trace types, cross-filtering, out-of-core, stateless server).

FlexViz separates axis locking from aggregation: a locked y-axis keeps its range, while line data re-aggregates only when the x viewport changes. Resetting restores the full viewport and fetches a fresh aggregate.

I would be very happy to help further if this workflow is still relevant.

**Reply by darynwhite** [reply]
Hi @jvdd, FlexViz looks like a very interesting a promising package. I'm curious if I could drop it in over plotly-resampler? 

This workflow is still very relevant, and looking at what FlexViz can do with Dashboards and such, I'm curious to use it in other applications as well. I've enjoyed Plotly Dash for uses and, assuming, that I can still customize plots and access the API underneath FlexViz for our purposes, I could see it being an easier place to start with our denser data sets.
